[asdfghjkxd] iP

georgetayqy opened the pull request:
# Juke 🤖

> The only Task Manager you need!

With Juke, you no longer need to worry about forgetting things! Just let Juke remember it for you!

Juke is, 

* Fully text-based and on the Command Line!
* ~easy~ SUPER EASY to learn!
* _EVEN EASIER to use!_ :0
* Works out of the box!
* Batteries-included! 
* **No pesky Pro/Paywalled features!!!! FREE OF CHARGE!!!!!**

To start using Juke,

1. Download the latest release of Juke from [this webpage](https://github.com/asdfghjkxd/ip/releases)
2. Save the JAR file to your computer
3. Open your PC's Terminal of choice and navigate to where the JAR file is downloaded, using the command `cd path/to/jar/file`, where `path/to/jar/file` represents the path to the downloaded JAR file
4. Execute the command: `java -jar juke.jar`
5. Sit back, and let Juke handle your tasks for you 😉

What Juke can do for you:

- [X] Managing different types of tasks: ToDo, Deadlines and Events
- [X] Intelligent Pattern Recognition Engines powered by Regex
- [X] Useful error messages to alert users of what mistakes they have done
- [X] Search and Query for tasks by name
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

Are you an aspiring Java Programmer? Use Juke _to practice your Java Skills_! Here's the `main` method for Juke:

```java
public static void main(String[] args) {
    try {
        Juke jukeAssistant = new Juke();
        jukeAssistant.start();
    } catch (JukeInitialisationException | JukeStorageException
             | JukeParseException ex) {
        // program should not continue if it cannot initialise properly
        // or if there are issues with retrieving data from the datafile
        new JukeExceptionCommand(ex).execute(Response.of());
    }
}
```

Have fun! 🎉🎉🎉

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

georgetayqy committed:
Implementation of Level-0 Requirements.

georgetayqy committed:
Implementation of Level-1 Requirements, along with improvements to Level-0 Requirements

georgetayqy committed:
Implementation of Level-2 Requirements

georgetayqy committed:
Implementation of Level-3 Requirements

georgetayqy committed:
Implementation of Level-4 Requirements

georgetayqy committed:
Implementation of Level-5 Requirements

georgetayqy committed:
Implementation of Level-6 Requirements

georgetayqy committed:
Implement Level-7 Requirements

georgetayqy committed:
Implement Level-7 Requirements

georgetayqy committed:
Implement Level-8 Requirements

georgetayqy committed:
Implement Level-8 Requirements

georgetayqy committed:
Organise Code into Respective Packages

georgetayqy committed:
Merge changes to Package Structure

georgetayqy committed:
Refactor classes and rearrange packages

georgetayqy committed:
Merge changes to classes and packages

georgetayqy committed:
Merge Gradle support into current repository

georgetayqy committed:
Update Gradle Configurations and Fix Style

georgetayqy committed:
Merge changes to Gradle files and Code Style

georgetayqy committed:
Implement Unit Tests

georgetayqy committed:
Merge implementation of Unit Tests

georgetayqy committed:
Implement JAR file Gradle builds

georgetayqy committed:
Merge JAR file Gradle builds

georgetayqy committed:
Update Javadoc comments

georgetayqy committed:
Amend Javadoc to comply with guidelines

georgetayqy committed:
Refactor code to meet Java coding standards

georgetayqy committed:
Implement Level-9 Requirements

georgetayqy committed:
Merge Level-9 changes

georgetayqy committed:
Merge improvements to Java coding standards

georgetayqy committed:
Merge changes made to Javadoc

georgetayqy committed:
Update JUnit tests to satisfy conventions

georgetayqy committed:
Abstract command handling logic

georgetayqy committed:
Merge changes to command handling logic

georgetayqy committed:
Enhance Juke with GUI

georgetayqy committed:
Refactor designs of UI into FXML stylesheets

georgetayqy committed:
Refactor code into new packages

georgetayqy committed:
Merge changes to the UI of the app

GlendaChong commented on src/main/java/juke/Juke.java:
```diff
@@ -0,0 +1,42 @@
+package juke;
+
+import javafx.application.Application;
+import juke.commands.JukeExceptionCommand;
+import juke.exceptions.JukeInitialisationException;
+import juke.exceptions.parsers.JukeParseException;
+import juke.exceptions.storage.JukeStorageException;
+import juke.responses.Response;
+import juke.ui.Ui;
```
I noticed the same issue for the other files. Shouldn't the imports follow the coding standard of Java with appropriate spacings?

georgetayqy replied:
Oh oops, I must have overlooked the import statements that the IDE injected into my code, will fix in a future commit!

GlendaChong commented on src/main/java/juke/commands/JukeCommand.java:
```diff
@@ -0,0 +1,262 @@
+            return JukeCommand.event(args, taskList);
+        case "find":
+            return JukeCommand.find(args, taskList);
+        default:
+            // exits the switch and throws an exception in the proceeding line
+            break;
+        }
+
+        // runs if no commands are matched
+        throw new JukeIllegalArgumentException("Oh no! I do not understand that command!");
```
Perhaps this throw statement could be placed in the default statement, instead of just having a break statement?

georgetayqy replied:
I had originally wanted the default action to be the throwing of the exception, and omitting the next throw statement, but for some reason, the linter detected this as an error! It is a little clunky for now, but I will try to see how I can fix this in a future commit!

GlendaChong commented on src/main/java/juke/commands/JukeCommand.java:
```diff
@@ -0,0 +1,262 @@
+                throw new JukeIllegalArgumentException("Oh no! The topic cannot contain the character \"|\"!");
+            }
+
+            LocalDateTime localTimeOne = DateTimeParser.parse(parsedArguments[1]);
+            LocalDateTime localTimeTwo = DateTimeParser.parse(parsedArguments[2]);
```
Perhaps a more intuitive variable name can be used here, to indicate start and end time?

georgetayqy replied:
I must have forgotten to change the names after rushing out the feature, will fix in a later commit!

GlendaChong commented on src/main/java/juke/commands/JukeCommand.java:
```diff
@@ -0,0 +1,262 @@
+     *
+     * @param response {@code Response} object that contains response from Juke and the user
+     * @return {@code Response} object that contains response from Juke and the user
+     */
```
Should you add @throws statement for this header comment?

georgetayqy replied:
Oh yikes, must have been another oversight. Thank you for noticing it!

GlendaChong commented on src/main/java/juke/commands/JukeFindTaskCommand.java:
```diff
@@ -0,0 +1,63 @@
+                    .append(this.word)
+                    .append("\":\n");
+
+            for (JukeTask t : foundTask) {
```
Should there be a space between "t" and ":" ?

georgetayqy replied:
I am currently following the standard coding convention as laid out in this Java Documentation: [link](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/for.html); it appears that leaving a space between the iterator variable and the variable to iterate through is the correct convention!

GlendaChong commented on src/main/java/juke/commons/utils/StringUtils.java:
```diff
@@ -0,0 +1,39 @@
+     * <p>
+     * Inspiration for this method comes from:
+     * <a href="https://stackoverflow.com/questions/41056517/breaking-a-long-url-to-several-line-in-javadoc">...</a>
+     *
```
I like how you provided the source reference for this method used!

GlendaChong commented on src/main/java/juke/tasks/JukeDeadline.java:
```diff
@@ -0,0 +1,67 @@
+     * @param completion Status of completion of the task
+     * @throws JukeStateException if the task is already completed
+     */
+    public JukeDeadline(String task, LocalDateTime deadline, boolean completion) {
```
Perhaps the name of the boolean term "completion" should follow the coding standard? This issue arises for a few boolean terms in other files too.

georgetayqy replied:
Ah, another one of my oversights when checking through my code. Thank you for reminding me!

GlendaChong commented on src/main/java/juke/tasks/TaskList.java:
```diff
@@ -0,0 +1,184 @@
+     */
+    public JukeTask deleteTask(int task) {
+        try {
+            JukeTask retTask = this.tasks.get(task);
```
Perhaps a more intuitive name could be used here instead of `retTask`?

georgetayqy replied:
Yup, I agree, will change the name in a future commit!

GlendaChong reviewed (commented):
Overall, I found your code easy to read for the most part! I noticed that your imports for some files are not following the coding standard we're using for this module. Just a minor issue on your naming of variables, but generally looks good!

thienmy0 commented on src/main/java/juke/commands/JukeCommand.java:
```diff
@@ -0,0 +1,262 @@
+        default:
+            // exits the switch and throws an exception in the proceeding line
+            break;
+        }
```
I like how neat your code is!

thienmy0 commented on src/main/java/juke/commands/JukeFindTaskCommand.java:
```diff
@@ -0,0 +1,63 @@
+     */
+    @Override
+    public Response execute(Response response) {
+        List<JukeTask> foundTask = this.taskList.findTask(this.word);
```
Perhaps this can be named foundTasks instead?

georgetayqy replied:
Oh right, another oversight on my part! Thank you for bringing this up!

thienmy0 commented on src/main/java/juke/parsers/DateTimeParser.java:
```diff
@@ -0,0 +1,86 @@
+        if (isValidDateTime(datetime)) {
+            String[] dateTimeSplit = datetime.split(" ");
+            String[] date = dateTimeSplit[0].split("\\/|-");
+            String[] time = Pattern.matches("\\d{4}", dateTimeSplit[1])
+                    ? new String[]{dateTimeSplit[1].substring(0, 2), dateTimeSplit[1].substring(2, 4)}
+                    : dateTimeSplit[1].split("-|:");
```
Perhaps this portion can be neater by defining the condition, consequent & alternative first so that it is less complicated
e.g.
```
boolean boolName = Pattern.matches("\\d{4}", dateTimeSplit[1]);
and so on
then
String[] time = boolName ? --- : ---
```

georgetayqy replied:
Ah I see, I will look into making this chunk of code more readable in a future commit!

thienmy0 reviewed (commented):
Your code looks neat overall!
It was pretty difficult to find issues, just some parts might be a little complicated.
Good job :+1:

georgetayqy committed:
Add assertions to ensure correctness of code

There are some methods that amend the underlying implementation
of the TaskList, and correctness (length of the list after amendments)
needs to be checked to ensure that invariants are not violated.

Adding assertions to the code improves code quality and correctness.

In an effort to ensure correctness, let's include some assertions at
crucial points of the code where correctness is important. Doing so
will improve code correctness.

georgetayqy committed:
Improve code quality of codebase

georgetayqy committed:
Merge pull request #1 from asdfghjkxd/branch-A-Assertions

Add assertions to ensure correctness of code

georgetayqy committed:
Merge branch 'master' into branch-A-CodeQuality

georgetayqy committed:
Merge pull request #2 from asdfghjkxd/branch-A-CodeQuality

Improve code quality of codebase

georgetayqy committed:
Implement CI with Github Actions

georgetayqy committed:
Merge pull request #3 from asdfghjkxd/branch-A-CI

Implement CI with Github Actions

XihuaZ reviewed (commented):
Coding standard is adhered to. Keep up the good work!

georgetayqy committed:
Implement Sort feature for tasks

Previously, the user does not have any reasonable way of sorting the
tasks that they have added other than to edit the datafile directly,
which may be dangerous as erroneous edits may cause the program to
crash.

Adding a sort feature ensures that the user is now able to better
organise their tasks in a ascending/descending order, based on the
task description, start date, end date and deadline of tasks.

To prevent user-induced errors due to careless edits of the datafile,
as well as to give users more functionality to customise their
task lists, let's give users a convenient but powerful sort feature
to allow them to do so.

georgetayqy committed:
Merge pull request #4 from asdfghjkxd/branch-C-Sort

Implement Sort feature for tasks

georgetayqy committed:
Fix bug with missing datafiles

georgetayqy committed:
Merge pull request #6 from asdfghjkxd/branch-Fix-Missing-Datafiles

Fix bug with missing datafiles

georgetayqy committed:
Add credits and code reuse comments

georgetayqy committed:
Merge pull request #8 from asdfghjkxd/branch-MoreCodeQuality

Add credits and code reuse comments

georgetayqy committed:
Merge branch 'master' into branch-MoreJUnit

georgetayqy committed:
Merge pull request #9 from asdfghjkxd/branch-MoreJUnit

Update JUnit tests to satisfy conventions

georgetayqy committed:
Fix code quality oversights

georgetayqy committed:
Merge pull request #10 from asdfghjkxd/branch-CodeQualityFixes

Fix code quality oversights

georgetayqy committed:
Add documentation for Github Pages

georgetayqy committed:
Fix minor bugs with GUI

georgetayqy committed:
Add more recognised parameters

georgetayqy committed:
Merge pull request #11 from asdfghjkxd/branch-A-UserGuide

Branch a user guide

georgetayqy committed:
Update GitHub Pages documentations

georgetayqy committed:
Merge pull request #12 from asdfghjkxd/branch-A-UserGuide

Update GitHub Pages documentations

georgetayqy committed:
Update Documentation

georgetayqy committed:
Merge pull request #13 from asdfghjkxd/branch-A-UserGuide

Update Documentation

georgetayqy committed:
Fix formatting errors in GitHub Pages

georgetayqy committed:
Merge pull request #14 from asdfghjkxd/branch-A-UserGuide

Fix formatting errors in GitHub Pages

georgetayqy committed:
Fix minor formatting errors

georgetayqy committed:
Merge pull request #15 from asdfghjkxd/branch-A-UserGuide

Fix minor formatting errors

georgetayqy committed:
Update documentation

georgetayqy committed:
Merge pull request #16 from asdfghjkxd/branch-A-UserGuide

Update documentation

georgetayqy committed:
Fix bugs and enhance UI

georgetayqy committed:
Merge pull request #17 from asdfghjkxd/branch-FinalFixes

Fix bugs and enhance UI

georgetayqy committed:
Update user guide

georgetayqy committed:
Merge pull request #18 from asdfghjkxd/branch-A-UserGuide

Update user guide

georgetayqy committed:
Update documentations

georgetayqy committed:
Merge pull request #19 from asdfghjkxd/branch-A-UserGuide

Update documentations

georgetayqy committed:
Enhance GUI and refactor files

georgetayqy committed:
Merge pull request #20 from asdfghjkxd/branch-A-BetterGui

Enhance GUI and refactor files

georgetayqy committed:
Update Ui images

georgetayqy committed:
Merge branch 'master' into branch-A-UserGuide

georgetayqy committed:
Merge pull request #21 from asdfghjkxd/branch-A-UserGuide

Update Ui images

georgetayqy committed:
Update GUI image

georgetayqy committed:
Merge pull request #22 from asdfghjkxd/branch-A-UserGuide

Update GUI image

georgetayqy committed:
Enhance UI Dialog Boxes

georgetayqy committed:
Update Documentations

georgetayqy committed:
Enhance code quality

georgetayqy committed:
Update JUnit test

georgetayqy committed:
Fix bug with sorting feature

georgetayqy committed:
Fix indexing bug in task list

georgetayqy committed:
Update documentation and error texts

georgetayqy committed:
Fix parser bugs

georgetayqy committed:
Fix bug with exit command

georgetayqy committed:
Fix checkstyle error

georgetayqy committed:
Merge pull request #24 from asdfghjkxd/branch-UiUpgrades

Branch UI upgrades

georgetayqy committed:
Test callout blocks

georgetayqy committed:
Merge pull request #25 from asdfghjkxd/branch-BetterUserGuide

Test callout blocks

georgetayqy committed:
Update callouts to Jekyll syntax

georgetayqy committed:
Merge pull request #26 from asdfghjkxd/branch-BetterUserGuide

Update callouts to Jekyll syntax

georgetayqy committed:
Rollback to previous docs

georgetayqy committed:
Merge pull request #27 from asdfghjkxd/branch-BetterUserGuide

Rollback to previous docs

georgetayqy committed:
Revert back to original version

georgetayqy committed:
Merge pull request #28 from asdfghjkxd/branch-BetterUserGuide

Revert back to original version

georgetayqy committed:
Fix errors in documentations

georgetayqy committed:
Merge pull request #29 from asdfghjkxd/branch-UiDocFixes

Fix errors in documentations

georgetayqy committed:
Update documentation

georgetayqy committed:
Merge pull request #30 from asdfghjkxd/branch-UiDocFixes

Update documentation

georgetayqy committed:
Fix excessive whitespace in dialog box

georgetayqy committed:
Merge pull request #31 from asdfghjkxd/branch-UiDocFixes

Fix excessive whitespace in dialog box

georgetayqy committed:
Update HTML code

georgetayqy committed:
Merge pull request #32 from asdfghjkxd/branch-UiDocFixes

Update HTML code

georgetayqy committed:
Remove expandable fields

georgetayqy committed:
Merge pull request #33 from asdfghjkxd/branch-UiDocFixes

Remove expandable fields

georgetayqy committed:
Add new unit tests for some tasks

georgetayqy committed:
Merge pull request #34 from asdfghjkxd/branch-MoreJUnitTests

Add new unit tests for some tasks

georgetayqy committed:
Fix bugs with sort feature

georgetayqy committed:
Merge pull request #35 from asdfghjkxd/branch-SortFixes

Fix bugs with sort feature

georgetayqy committed:
Fix typo

georgetayqy committed:
Merge pull request #36 from asdfghjkxd/branch-SortFixes

Fix typo

georgetayqy committed:
Fix documentation

georgetayqy committed:
Merge pull request #37 from asdfghjkxd/branch-SortFixes

Fix documentation

georgetayqy committed:
Remove Data Files

georgetayqy committed:
Merge pull request #39 from asdfghjkxd/branch-Remove-Data-Files

Remove Data Files
